### ISSUE TITLE
SSH Agent - Public Key Denied

### ISSUE THREAD

**Issue by CodeMan99** [microsoft/vscode-remote-release#8666]
I followed the [Sharing git credentials](https://code.visualstudio.com/remote/advancedcontainers/sharing-git-credentials#_using-ssh-keys) resource to setup my windows _ssh agent_. I believe it worked at first, but now my public key is being rejected when trying to use git commands in the container (or any ssh connection).

I have verified that `ssh-add -l` reports the same fingerprint in all environments. The same exact ssh key is working fine in both windows and WSL (Ubuntu). Only in my devcontainer is the key being rejected.

---

- VSCode Version: `1.79.2 695af097c7bd098fbf017ce3ac85e09bbc5dda06 x64`
- Local OS Version: `Windows 11 Home 22H2 build 22621.1848`
- Remote OS Version: `Debian GNU/Linux 12 (bookworm)`
- Remote Extension/Connection Type: `Containers`
- Image: `mcr.microsoft.com/devcontainers/javascript-node:1-18-bookworm`

- Logs:

```
$ git remote update
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Steps to Reproduce:

Currently unknown.

Does this issue occur when you try this locally?: No
Does this issue occur when you try this locally and all extensions are disabled?: Unknown

**Comment by CodeMan99** [comment]
Huh. Seems that both `:0-18-bullseye` and `:1-18-bullseye` are working ok. The issue seems to be with the _bookworm_ image.

**Comment by chrmarti** [comment]
Using the bookworm image with `git@github.com:microsoft/vscode-remote-try-node.git` works for me.

Please append the Dev Containers log from when this happens. (`F1` > `Dev Containers: Show Container Log`)

**Comment by CodeMan99** [comment]
```
[6491 ms] Start: Run in container: mkdir -p '/home/node/.vscode-server/bin' && ln -snf '/vscode/vscode-server/bin/linux-x64/695af097c7bd098fbf017ce3ac85e09bbc5dda06' '/home/node/.vscode-server/bin/695af097c7bd098fbf017ce3ac85e09bbc5dda06'
[6495 ms] 
[6495 ms] 
[6495 ms] Start: Run in Host: docker exec -i -u root ba6841e7dd24530ab214754987fd20e4b234bb72d37dcfd91aa8a3cdababba3a /bin/sh
[6495 ms] Start: Launching Dev Containers helper.
[6496 ms] ssh-agent: SSH_AUTH_SOCK not set on wsl host.
[6496 ms] ssh-agent: SSH_AUTH_SOCK not set on local host.
[6496 ms] ssh-agent: SSH_AUTH_SOCK in container (/tmp/vscode-ssh-auth-c574dd6f-c5eb-4593-a1f2-919ed41d1148.sock) forwarded to local host (\\.\pipe\openssh-ssh-agent).
[6496 ms] Start: Run in container: test -e /tmp/.X11-unix/X0
[6499 ms] 
[6499 ms] 
[6499 ms] Exit code 1
```

Again, the fingerprint show up fine with `ssh-add -l`.

**Comment by CodeMan99** [comment]
The log is identical on bullseye.

**Comment by chrmarti** [comment]
What version do you get for `ssh -V` locally on Windows? If that is 8.8 or earlier, try upgrading to 8.9 or later (there are MSI installers available here: https://github.com/PowerShell/Win32-OpenSSH/releases).

This might be https://github.com/microsoft/vscode-remote-release/issues/7525#issuecomment-1354742843, bullseye is at ssh 8.4 and bookworm is at 9.2.

**Comment by hojo0590** [comment]
I'm experiencing exactly the same. `ssh-add -l` correctly shows ssh keys imported to the host's `ssh-agent` in all cases, I'm using Windows11 Build 22621.1848 as host with `ssh -V` reporting as `OpenSSH_for_Windows_8.6p1, LibreSSL 3.4.3`.
If I run a minimal devcontainer
```json
{
	"name": "Ubuntu",
	"image": "mcr.microsoft.com/devcontainers/base:focal"
}
```
which contains `ssh -V` `OpenSSH_8.2p1 Ubuntu-4ubuntu0.7, OpenSSL 1.1.1f  31 Mar 2020` a connection to `github`
```bash
ssh -v git@github.com
```
will succeed.

When replacing the container image to `mcr.microsoft.com/devcontainers/base:jammy` (20.04 -> 22.04) which contains  `OpenSSH_8.9p1 Ubuntu-3ubuntu0.1, OpenSSL 3.0.2 15 Mar 2022` the command
```bash
ssh -v git@github.com
```
will fail and return `git@github.com: Permission denied (publickey)`

Temporary fix for me (as I can't update the OpenSSH on the host) is to just use the `focal` base image.

**Comment by hojo0590** [comment]
@CodeMan99 I have been trying to fix this for quite some while - if you have a look at [Devcontainer Docs](https://code.visualstudio.com/docs/devcontainers/containers#_dev-containers-limitations) all the information is there. I wasn't able to stumble upon that before finding this issue.
@chrmarti as more recent base images surely will increase in popularity my two cents on that documentation: it would have saved me a lot of time if that limitation also stated that `ssh-add -l` will correctly show the key but any git action will result in `Permission denied (publickey).`.

**Comment by CodeMan99** [comment]
@chrmarti My windows SSH version is:

```
PS C:\> ssh -V
OpenSSH_for_Windows_8.6p1, LibreSSL 3.4.3
```

I'll attempt the SSH upgrade later this week. I just returned from vacation.

**Comment by CodeMan99** [comment]
@chrmarti Yes, the upgrade to OpenSSH worked fine with the bookworm image.

```
PS C:\> ssh -V
OpenSSH_for_Windows_9.2p1, LibreSSL 3.7.2
```

I'll add that I upgraded via ` winget install --id Microsoft.OpenSSH.Beta --override ADDLOCAL=Client` (from this [superuser](https://superuser.com/a/1722263) answer).

**Comment by CodeMan99** [comment]
I would personally consider this complete if there's something in the docs and/or logs of the devcontainer indicating to upgrade OpenSSH.

**Comment by chrmarti** [comment]
@CodeMan99 We have added a note to https://code.visualstudio.com/docs/devcontainers/containers#_dev-containers-limitations.

Keeping this issue open for adding a runtime check. Thanks.

**Comment by sketch34** [comment]
I am encountering a very similar issue since a couple of days ago. `ssh-agent` is not running on the Docker host (bookworm) and `git@github.com: Permission denied (publickey)` occurs after vscode devcontainers connects. SSH versions are higher than this issue mentions though? I'm not sure what has introduced the issue, the same setup has been running fine until recently.

```
[14165 ms] Start: Launching Dev Containers helper.
[14166 ms] ssh-agent: SSH_AUTH_SOCK not set on wsl host.
[14166 ms] ssh-agent: SSH_AUTH_SOCK not set on local host.
[14166 ms] Start: Run in container: test -e /tmp/.X11-unix/X0
[14169 ms] 
[14169 ms] 
[14169 ms] Start: Run in container: test -e /tmp/.X11-unix/X1
[14171 ms] 
[14171 ms] 
[14171 ms] Exit code 1
[14171 ms] Start: Run in container: mkdir -p '/tmp/.X11-unix'
[14174 ms] 
[14174 ms] 
[14174 ms] X11 forwarding: DISPLAY in container (:1) forwarded to wsl host (:0).
[14175 ms] Start: Run in container: gpgconf --list-dirs
[14180 ms] sysconfdir:/etc/gnupg
```

```
C:\Windows\System32\OpenSSH>ssh -V
OpenSSH_for_Windows_9.5p1, LibreSSL 3.8.2
```

```
appuser@docker-desktop:~/app$ ssh -V
OpenSSH_9.2p1 Debian-2+deb12u6, OpenSSL 3.0.16 11 Feb 2025
```

**Comment by sketch34** [comment]
Sorry for noise but perhaps this helps someone.

Firstly the [sharing git credentials](https://code.visualstudio.com/remote/advancedcontainers/sharing-git-credentials) documentation is ambiguous on what is meant by "the host". In my case I am on Windows 11 with a Docker container running off WSL2. So it took me some trial and error to discover "the host" in this context is the WSL2 distro. No changes were required in Windows.

I added the `if [ -z "$SSH_AUTH_SOCK" ]; then ...` to my `~/.profile` and git connectivity was restored in the container. 

The steps concerning agent forwarding seem to not be required. `sshd_config` is not present for me in Windows or WSL2 distro.

The big question for me is how it was working fine for weeks and then stopped ~3 days ago. It seemed to correlate with the recent Windows update to 10.0.26100.

**Comment by chrmarti** [comment]
Maybe the `OpenSSH Authentication Agent` service is no longer enabled on Windows?

**Comment by JohnStrunk** [comment]
I too have just run into this. I share my yubikey via gpg-agent to use with git and ssh in the devcontainer... It used to work fine, but has recently stopped on my main Windows 11 machine. After checking on another Win11 machine today, it worked w/ vscode 1.96.2, but when the auto-update kicked in and moved that machine to 1.101.1, it's now broken.

EDIT: My issue is #11043 , and the workaround here fixed it: https://github.com/microsoft/vscode-remote-release/issues/11043#issuecomment-3005677524
